src/Data/Machine/Group.hs

```diff
@@ -1,59 +1,162 @@
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE GADTs #-}
+
+-- | Split up input streams into groups with separator values and process the
+-- groups with their own 'MachineT'.
+
 module Data.Machine.Group
   ( groupingOn
+  , groupingOn_
+    -- * Tagging a stream
   , taggedBy
+  , taggedBy_
+  , taggedAt
+  , taggedAt_
+  , taggedAtM
+  , taggedCount
+    -- * Reset a machine for each group
   , partitioning
+  , partitioning_
+    -- * Helpers
   , starve
   , awaitUntil
-  )where
+  ) where
+
+import Control.Monad (guard)
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.State.Strict (get, put)
 import Data.Machine
+import Data.Machine.Lift (execStateM)
+
+#if __GLASGOW_HASKELL__ < 710
+-- $setup
+-- >>> import Control.Monad.Trans.Reader (ask, runReader)
+-- >>> import Control.Applicative ((<$))
+#else
+-- $setup
+-- >>> import Control.Monad.Trans.Reader (ask, runReader)
+#endif
 
 isLeft :: Either a b -> Bool
 isLeft = either (const True) (const False)
 
--- | Using a function to signal group changes, apply a machine independently over each group.
-groupingOn :: Monad m => (a -> a -> Bool) -> ProcessT m a b -> ProcessT m a b
-groupingOn f m = taggedBy f ~> partitioning m
+-- | Using a function to signal group changes, apply a machine independently
+-- over each group.
+groupingOn_ :: Monad m => (a -> a -> Bool) -> ProcessT m a b -> ProcessT m a b
+groupingOn_ f m = taggedBy_ f ~> partitioning_ m
+{-# INLINE groupingOn_ #-}
+
+-- | Using a function to signal group changes, apply a machine independently
+-- over each group using the difference between the group edges.
+groupingOn :: Monad m => i -> (a -> a -> Maybe i) -> (i -> ProcessT m a b) -> ProcessT m a b
```

Comment:
What does "using the difference between the group edges" mean?

Reply:
I was thinking where two items materially differ it would return the difference in the Just.  It could be `edge value` perhaps?  I wasn't sure how to describe it and hoped having a crappy comment to improve would be better than nothing.